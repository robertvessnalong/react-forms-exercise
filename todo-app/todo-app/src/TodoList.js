import { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';

const TodoList = () => {
  const INITAL_STATE = [];

  const [todos, setTodo] = useState(INITAL_STATE);

  const addTodo = (newTodo) => {
    setTodo((todos) => [...todos, { ...newTodo, id: uuidv4() }]);
  };

  const updateTodo = (id, newTodo) => {
    setTodo((todos) =>
      todos.map((todo) => (todo.id === id ? { ...todo, todo: newTodo } : todo))
    );
  };

  const removeTodo = (id) => {
    const newTodos = todos.filter((todo) => todo.id !== id);
    setTodo(newTodos);
  };

  return (
    <div>
      <h3>Todo List</h3>
      <NewTodoForm addTodo={addTodo} />
      <div className='todos'>
        {todos.map((todo, idx) => (
          <Todo
            id={todo.id}
            todo={todo.todo}
            key={todo.id}
            idx={idx}
            remove={removeTodo}
            edit={updateTodo}
          />
        ))}
      </div>
    </div>
  );
};

export default TodoList;
