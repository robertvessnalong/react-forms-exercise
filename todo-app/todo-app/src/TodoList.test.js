import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

it('renders without crashing', () => {
  render(<TodoList />);
});

it('matches snapshot', () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it('add new item', () => {
  const { getByLabelText, getByText } = render(<TodoList />);
  const input = getByLabelText('Add Todo:');
  const btn = getByText('Submit');
  fireEvent.change(input, { target: { value: 'Go to Sleep' } });
  fireEvent.click(btn);

  expect(input).toHaveValue('');
  expect(getByText('1 - Go to Sleep')).toBeInTheDocument();
  expect(getByText('Edit')).toBeInTheDocument();
  expect(getByText('X')).toBeInTheDocument();
});

it('edit todo', () => {
  const { getByLabelText, getByText, getByDisplayValue, queryByText } = render(
    <TodoList />
  );
  const input = getByLabelText('Add Todo:');
  const btn = getByText('Submit');
  fireEvent.change(input, { target: { value: 'Go to Sleep' } });
  fireEvent.click(btn);
  fireEvent.click(getByText('Edit'));
  const edit = getByDisplayValue('Go to Sleep');
  fireEvent.change(edit, { target: { value: 'Wake Up' } });
  fireEvent.click(getByText('Edit Todo'));

  expect(getByText('1 - Wake Up')).toBeInTheDocument();
  expect(queryByText('1 - Go to Sleep')).not.toBeInTheDocument();
});

it('can delete a todo', () => {
  const { getByLabelText, getByText, queryByText } = render(<TodoList />);
  const input = getByLabelText('Add Todo:');
  const btn = getByText('Submit');
  fireEvent.change(input, { target: { value: 'Go to Sleep' } });
  fireEvent.click(btn);

  fireEvent.click(getByText('X'));

  expect(queryByText('1 - Go to Sleep')).not.toBeInTheDocument();
});
