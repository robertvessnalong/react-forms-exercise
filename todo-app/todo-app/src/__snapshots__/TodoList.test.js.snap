// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Todo List
    </h3>
    <form>
      <label
        for="todo"
      >
        Add Todo:
      </label>
      <input
        id="todo"
        name="todo"
        placeholder="Todo"
        type="text"
        value=""
      />
      <button>
        Submit
      </button>
    </form>
    <div
      class="todos"
    />
  </div>
</DocumentFragment>
`;
